qiskit/circuit/add_control.py

```diff
@@ -14,22 +14,31 @@
 """
 Add control to operation if supported.
 """
-from qiskit import QiskitError
+from typing import Union, Optional
+
+from qiskit.circuit.exceptions import CircuitError
 from qiskit.extensions import UnitaryGate
+from . import ControlledGate, Gate, QuantumRegister, QuantumCircuit
 
 
-def add_control(operation, num_ctrl_qubits, label):
-    """Add num_ctrl_qubits controls to operation
+def add_control(operation: Union[Gate, ControlledGate],
+                num_ctrl_qubits: int,
+                label: Union[str, None],
+                ctrl_state: Union[int, str, None]) -> ControlledGate:
+    """Adds num_ctrl_qubits controls to operation.
```

Comment:
Can you add a bit more documentation here, describing how the controls are done?
For example: 
```
For standard gates, if the controlled version already exists in the library, it will be returned (e.g. XGate.control() = CnotGate().

For more generic gates, the controlled version is implemented by first decomposing into a ["u1", "u3", "cx"] basis, then controlling each gate in the decomposition.

Open controls are implemented by conjugating the control line with X gates.
```

Reply:
ok